extensions/amp-video/0.1/test-e2e/test-amp-video-flexible-bitrate.js

```diff
@@ -0,0 +1,232 @@
+/**
+ * Copyright 2021 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+const VIEWPORT = {
+  HEIGHT: 768,
+  WIDTH: 1024,
+};
+
+const VIDEO_EVENTS = {
+  DOWNGRADE: 'd',
+  LOAD: 'l',
+  UNLOAD: 'u',
+};
+
+describes.endtoend(
+  'amp-video flexible bitrate',
+  {
+    fixture: 'amp-video/videos-cdn.html',
+    environments: ['single'],
+    experiments: ['flexible-bitrate'],
+    initialRect: {width: VIEWPORT.WIDTH, height: VIEWPORT.HEIGHT},
+  },
+  (env) => {
+    let controller;
+    let story;
+    let debugField;
+
+    beforeEach(async () => {
+      controller = env.controller;
+
+      story = await controller.findElement('amp-story.i-amphtml-story-loaded');
+      await controller.findElement('amp-story-page#page-1[active]');
+
+      debugField = await controller.findElement('input#flexible-bitrate-debug');
+    });
+
+    /**
+     * Tests for first load
+     */
+
+    it('when loading the page with good connection, the best quality source is played', async () => {
+      const video1El = await controller.findElement('#video1 video');
+      await expect(
+        await controller.getElementProperty(video1El, 'currentSrc')
+      ).contains('#high');
+    });
+
+    it("when on the first page, a further video doesn't load", async () => {
+      const video4El = await controller.findElement('#video4 video');
+      await expect(
+        await controller.getElementProperty(video4El, 'currentSrc')
+      ).to.equal('');
+    });
+
+    /**
+     * Tests for downgrades on active video
+     */
+
+    it('when a video buffers, the video changes to a lower quality source', async () => {
+      await forceEventOnVideo(VIDEO_EVENTS.UNLOAD, 1);
+
+      const video1El = await controller.findElement('#video1 video');
+
+      await forceEventOnVideo(VIDEO_EVENTS.DOWNGRADE, 1);
+
+      await expect(
+        await controller.getElementProperty(video1El, 'currentSrc')
+      ).contains('#med');
+    });
+
+    it('when a video buffers, the low quality source starts playing at the same place the previous source was stopped at', async () => {
```

Comment:
When you split a test in two so you keep one assertion per test (very good practice 👍 ) you need to keep the exact same conditions. Please make sure you have the exact same instructions and instructions order, that makes it easier to spot any difference when debugging or reading the code.

Reply:
This test has a delay of 100ms so that the `currentTime` specifically advances a bit before downgrading (which is unique to this test). I'm updating the rest of the test to look more similar otherwise